058-kraft-upgrades-and-downgrades.md

```diff
@@ -0,0 +1,228 @@
+As a result, the first step of the downgrade procedure - downgrading the `metadata.version` - can be done safely without any metadata loss only for selected metadata versions.
+For other versions, the downgrade will be possible only as an _unsafe_ downgrade that might result in losing some metadata information, which might have a negative impact on the cluster.
+This is similar to how the `inter.broker.protocol.version` works today, as it also doesn't offer any guarantees of full compatibility when downgrading it.
+
+#### Existing downgrade limitations
+
+As of today, the Kraft downgrade situation is as follows:
+1. The unsafe downgrade is currently not supported by Apache Kafka.
+2. None of the `metadata.versions` supported by current Kafka versions support the safe downgrade.
+   _(You should be able to downgrade from 3.5-IV1 to 3.5-IV0. But you cannot downgrade between 3.5 and 3.4 metadata versions.)_
+
+The first point should be addressed in the future when the support for the unsafe downgrade is implemented.
+The second point is expected to remain an issue.
+While in the future the metadata formats might become more stable and it might be possible to safely downgrade the metadata between different Kafka versions, there still might be some versions that do not support the safe downgrade.
+
+Despite Kafka currently not really supporting downgrades, this proposal designs the Strimzi implementation based on how the downgrade is expected to work.
+
+### Strimzi implementation
+
+Strimzi will follow the same approach to upgrading / downgrading a KRaft cluster as it does today for ZooKeeper-based clusters.
+
+#### Configuring the metadata version
+
+In ZooKeeper-based clusters, the `inter.broker.protocol.version` is stored in the Kafka configuration (`Kafka.spec.kafka.config`).
+The `metadata.version` is not part of the Kafka configuration.
+To allow users to configure it in our Kafka custom resource, a new String-type field named `metadataVersion` will be added to `Kafka.spec.kafka`.
+This field will be validated by the operator to contain a valid Kafka metadata version.
+
+Users can use this new field to configure the metadata version.
+The default value of this field when not set will be the metadata version corresponding to the current Kafka version.
+The default version corresponding to a given Kafka version will be stored in `kafka-versions.yaml` in the same way we today store the default `inter.broker.protocol.version` and `log.message.format.version`.
+
+The operator will use this field to set the metadata version in the Kafka cluster.
+In the initial deployment of the Kafka cluster, it will be set using the `kafka-storage.sh` tool.
+In existing Kafka clusters, it will be queried and changed using the Kafka Admin API.
+
+The current metadata version will be tracked in the `.status` section of the Kafka CR in `.status.kafkaMetadataVersion`.
```

Comment:
We are going to have `status.metadataState` for the ZooKeeper to KRaft migration FSM.
I was wondering if ... we should have `status.metadataVersion` here or the other way around changing on the migration side by having `status.kafkaMetadataState`. I have no strong opinion about it tbh.
Of course, these names make the two status fields related somehow. If I was a user I could think "my metadata state is .... " and "its version is". I think that's correct, because when metadata state is ZooKeeper I don't expect to be a metadata version there. When the cluster is KRaft both fields will be set. What about during a migration? I guess the metadata version should be set at least at the end of the migration process or even before during the dual write? Maybe the latter, because in dual write we are effectively using KRaft as metadata store (while copying to ZooKeeper as well).

Reply:
I opted for `kafkaMetadataVersion` because we already have `kafkaVersion`. I'm fairly flexible. I can rename it `metadataVersion` if you prefer. But we can't rename `kafkaVersion` anymore, so that is set.

Reply:
Well it comes to try thinking if "metadata" could be related to something else and not just "Kafka metadata".
I would say, let me change my `status.metadataState` to `status.kafkaMetadataState` so we are also aligned better with the `kafkaVersion` field.

What about my last question on setting the metadata version during migration?

Reply:
> What about during a migration? I guess the metadata version should be set at least at the end of the migration process or even before during the dual write? Maybe the latter, because in dual write we are effectively using KRaft as metadata store (while copying to ZooKeeper as well).

In the status, it should be IMHO set from the point when the controllers are enabled. For the documentation, we should also consider suggesting the users to set the `Kafka.spec.kafka.metadataVersion` field explicitely in the Kafka CR.

But I do not expect any major impact on the code. Depending on which Pr is merged first, the other would need to adjust the `isKraftEnabled` boolean flag versus the migration state. But apart from that, you should not need to write anything special in the migration code to set the metadata version or anything like that.

Reply:
I'm not sure if you want to include this in the proposal or not. I think it is a good point, b ut I would probably prefer to not include the details in it since they are a bit unclear and depend on the exact code (that is I guess still WiP on your side and does not exist for this proposal).